paderbox/utils/nested.py

```diff
@@ -400,3 +413,295 @@ def squeeze_nested(orig):
+            the nested container (and cannot be broadcasted if `broadcast=True`)
+
+    Returns:
+        Value located at `path` in `container`
+
+    Examples:
+        >>> d = {'a': 'b', 'c': {'d': {'e': 'f'}, 'g': [1, [2, 3], 4]}}
+        >>> get_by_path(d, 'a')
+        'b'
+        >>> get_by_path(d, 'c.d.e')
+        'f'
+        >>> get_by_path(d, ('c', 'g', 1, 0))
+        2
+        >>> get_by_path(d, 'a.b.c', allow_early_stopping=True)
+        'b'
+        >>> get_by_path(d, 'c.b.c', default=42)
+        42
+    """
+    if path is None:
+        return container
+
+    if isinstance(path, str):
+        path = path.split(sep)
+
+    for k in path:
+        try:
+            container = container[k]
+        except Exception as e:
+            # Indexing a custom type can raise any exception, in which case
+            # we try to broadcast
+            # Not sure if broadcasting makes sense for lists/tuples. It is
+            # hard to check for custom sequences because of str, so
+            # sequences are broadcasted here
+            if allow_early_stopping and not isinstance(container, Mapping):
+                return container
+
+            # We can't add another more specific except block because we have to
+            # catch all exceptions for the broadcasting. Assuming here that
+            # custom containers raise KeyErrors and IndexErrors correctly when
+            # the indexed element is not found
+            if isinstance(e, (KeyError, IndexError)):
+                if default is not ...:
+                    return default
+
+            raise
+    return container
+
+
+def set_by_path(
+        container: Union[Mapping, Sequence],
+        path: Union[str, Tuple[Any, ...], None],
+        value: Any,
+        *,
+        sep: str = '.',
+) -> None:
+    """
+    Sets a value in the nested dictionary `d` by the dotted path.
+
+    Modifies `d` inplace.
+
+    Args:
+        container: The container to get the value from
+        path: Dotted path or tuple of keys to index the nested container with.
+            A `tuple` is useful if not all keys are strings. If it is a `str`,
+            keys are delimited by `delimiter`
+        value: The value to set in `d` for `path`
+        sep: The delimiter for keys in path
+
+    Examples:
+        >>> d = {}
+        >>> set_by_path(container, 'a', {})
+        >>> container
+        {'a': {}}
+        >>> set_by_path(container, 'a.b', {'c': [1, 2, 3], 'd': 'e'})
+        >>> container
+        {'a': {'b': {'c': [1, 2, 3], 'd': 'e'}}}
+        >>> set_by_path(container, ('a', 'b', 'c', 2), 42)
+        >>> container
+        {'a': {'b': {'c': [1, 2, 42], 'd': 'e'}}}
+    """
+    if isinstance(path, str):
+        path = path.split(sep)
+    container = get_by_path(container, path[:-1])
+    container[path[-1]] = value
+
+
+def nested_iter_items(container, *, iter_types=(dict, tuple, list)):
```

Comment:
Maybe adding a comment, that this function could get a `depth` or `max_depth` argument in the future, when someone needs it??